x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/saml/SamlRedirectTests.java

```diff
@@ -64,53 +72,99 @@ public void testRedirectUrlWithRelayStateAndSigning() throws Exception {
-                    + "1yHnLOX4Edtz0eDuf2uJjHy9Z%2Fl5ZEapyLQvGraBZdZm6ER59RV%2F8izGKwLg38VL4B1sji%2FPxxgeIb"
-            )
+            "<?xml version=\"1.0\" encoding=\"UTF-8\"?>"
+                + "<saml2p:LogoutRequest xmlns:saml2p=\"urn:oasis:names:tc:SAML:2.0:protocol\" "
+                + "Destination=\"https://idp.test/saml/logout?\" "
+                + "ID=\"_id123456789\" IssueInstant=\"2018-01-14T22:47:00.000Z\" Version=\"2.0\">"
+                + "<saml2:Issuer xmlns:saml2=\"urn:oasis:names:tc:SAML:2.0:assertion\">https://idp.test/</saml2:Issuer>"
+                + "<saml2:NameID xmlns:saml2=\"urn:oasis:names:tc:SAML:2.0:assertion\">name-123456-7890</saml2:NameID>"
+                + "</saml2p:LogoutRequest>",
+            emptyMap(),
+            false
         );
     }
 
+    private void assertRedirectUrl(
+        String actualRequestUrl,
+        String expectedUncompressedSamlRequestContent,
+        Map<String, String> expectedParams,
+        boolean allowExtraParams
+    ) {
+        Set<String> requiredParams = new HashSet<>(expectedParams.keySet());
+        String[] parts = actualRequestUrl.split("\\?");
+        assertThat(parts[0], equalTo(LOGOUT_URL));
+        final String[] params = parts[1].split("&");
+        for (String param : params) {
+            final String[] keyValue = param.split("=", 2);
+            assertThat(keyValue.length, equalTo(2));
+            final String actualKey = keyValue[0];
+            final String actualValue = keyValue[1];
+            if (actualKey.equals("SAMLRequest")) {
+                final String samlRequest = URLDecoder.decode(actualValue, StandardCharsets.UTF_8);
+                final String decompressed = decompressAndBase64Decode(samlRequest);
+                assertThat(decompressed, equalTo(expectedUncompressedSamlRequestContent));
+            } else {
+                if (expectedParams.containsKey(actualKey)) {
+                    assertThat("invalid value for key " + actualKey, actualValue, equalTo(expectedParams.get(actualKey)));
+                    requiredParams.remove(actualKey);
+                } else if (allowExtraParams == false) {
+                    fail("unexpected parameter: " + actualKey);
+                }
+            }
+        }
+        assertThat(requiredParams, empty());
+    }
+
+    private String decompressAndBase64Decode(String compressedBase64) {
+        byte[] decodedBytes = Base64.getDecoder().decode(compressedBase64);
+        Inflater inflater = new Inflater(true);
+        inflater.setInput(decodedBytes);
+        byte[] result = new byte[decodedBytes.length * 10];
+        try {
+            int resultLength = inflater.inflate(result);
+            inflater.end();
+            return new String(result, 0, resultLength, StandardCharsets.UTF_8);
+        } catch (DataFormatException e) {
+            throw new RuntimeException("Failed to decompress SAML request", e);
+        }
+    }
```

Comment:
It could be simplified to avoid dealing with output byte array.
```suggestion
    private static String decompressAndBase64Decode(String compressedBase64) {
        byte[] compressed = Base64.getDecoder().decode(compressedBase64);
        Inflater inflater = new Inflater(true);
        try (InflaterInputStream in = new InflaterInputStream(new ByteArrayInputStream(compressed), inflater)) {
            return new String(in.readAllBytes(), StandardCharsets.UTF_8);

        } catch (IOException e) {
            throw new UncheckedIOException("Failed to decompress input: " + compressedBase64, e);
        }
    }
```